Add support for triage and maintain permissions

bluekeyes opened the pull request:
Deprecate the `admins` and `write_collaborators` properties and replace them with a new `permissions` key that accepts all permission levels. Like before, users must specify levels explicitly and there is no permission inheritance (e.g. users with admin permissions do not count as having write permissions.)

The GitHub REST API does not expose the new permissions, so switch collaborator lookup to GraphQL, which does include them. This will remove caching of collaborator information, so we may need to revisit that to reduce request rates.

I still need to fix up reviewer assignment to work with the new fields. That's coming in a future commit.

bluekeyes committed:
Add support for triage and maintain permissions

Deprecate the "admins" and "write_collaborators" properties and replace
them with a new "permissions" key that accepts all permission levels.
Like before, users must specify levels explicitly and there is no
permission inheritance (e.g. users with admin permissions do not count
as having write permissions.)

The GitHub REST API does not expose the new permissions, so switch
collaborator lookup to GraphQL, which does include them. This will
remove caching of collaborator information, so we may need to revisit
that to reduce request rates.

bluekeyes commented on pull/permission.go:
```diff
@@ -0,0 +1,51 @@
+	"strings"
+)
+
+type RepositoryPermission uint8
```
I'm unsure about this permission type. Using a number means we can easily check if a given permission includes another one, for example, that people with admin also have write. On the other hand, I don't use that ability anywhere in the current code and it means we need extra functions to parse and generate string representation.

But some "parsing" might be needed even if using string values to validate that inputs are valid permissions...

asvoboda replied:
I left another comment around this, but we currently have some existing string normalization. I don't feel strongly about internally representing these as numbers or strings, but I think we should be consistent about it.

bluekeyes replied:
I decided to stick with integers and ended up using the permission level comparison.

bluekeyes commented on policy/common/actor.go:
```diff
@@ -30,19 +30,23 @@ type Actors struct {
+	// Each desired permission must be listed explicity. For example, even
+	// though "admin" is a superset of "write" in GitHub, both "admin" and
+	// "write" must be included in the list to allow users with either role.
+	Permissions []string
```
I also considered a single `Permission` field that specifies a minimum permission. I think that's nicer in someways, but we need the list behavior to replace the `Admins` and `WriteCollaborators` fields, which worked with exact permissions rather than minimum permissions.

For approval, I'm not sure why you'd want something other than minimum permissions, but I can see some value in only assigning reviews to maintainers and not admins. Unfortunately, both approval and reviewer assignment use the same configuration.

We could also have two fields to support both behaviors, but that seemed too complicated.

asvoboda replied:
I think specifying the exact permissions is less of a possible foot-gun than a "minimum" permission level. 

On your second point, IIRC separating admin (vs admin via org owner) vs write was one of the topics around the reviewer assignment feature. I think it gets kind of cludgey and awkward, but we could introduce a new assignment mode for `maintainer but not admin`?

bluekeyes replied:
I ended up implementing a direct permission check: the user or team must be granted an exact permission specified in the list by the repository. This matches current behavior for `admins` and changes the behavior for `write_collaborators` to be consistent with `admins`.

It also means there's no way to request reviews from organization owners by using permissions, unless they are also part of a team or added directly. Since we didn't support this before, it should be fine to continue not supporting it.

bluekeyes committed:
Move CollaboratorPermissions to pull.Context

This will always operate on the current pull request/repository and can
share cached data with RepositoryCollaborators() so there's no reason to
make it part of the membership/cross-org context.

asvoboda commented on pull/github.go:
```diff
@@ -336,6 +336,31 @@ func (ghc *GitHubContext) RepositoryCollaborators() (map[string]string, error) {
+	return PermissionNone, nil
+}
+
 func coalescePermission(perms map[string]bool) string {
```
Do we also need to update `coalescePermission` to standardize the new triage/maintainer permissions? Alternatively, should we replace this func with the new `RepositoryPermission` type?

bluekeyes replied:
This function is removed in a future commit that fixes up reviewer assignment and converts `RepositoryCollaborators` to use the new permission type.

asvoboda commented on README.md:
```diff
@@ -324,10 +324,20 @@ requires:
+  # A user must have a permission in this list for their approval to count for
+  # this rule. Valid permissions are "admin", "maintain", "write", "triage",
+  # and "read".
+  permissions: ["admin", "maintain", "write"]
```
💯  this is much nicer.

bluekeyes committed:
Rename permission type, use it directly in config

Rename RepositoryPermission to Permission and use it directly in the
YAML configuration. This avoids parsing later on and adds validation,
since the current validation logic really only checks that the YAML
parses cleanly.

bluekeyes committed:
Return new permissions for teams/collaborators

This required quite a bit of API fiddling because the APIs here are
pretty inconsistent. I settled on a heuristic approach, but an alternate
method would be to make more API calls so that we:

1. List team permissions and then expand membership
2. List direct repository collaborators
3. Anything else must come from the org

bluekeyes committed:
Update reviewer assignment to use new permissions

This requires fixing up some tests which happened to work before because
write_collaborators did not work for teams. Also remove a test that was
only testing the mock pull.Context implementation.

bluekeyes committed:
Add license headers to new files

bluekeyes commented on pull/github.go:
```diff
@@ -927,6 +976,86 @@ func (ga v4GitActor) GetV3Login() string {
+
+type v4PermissionSources []v4PermissionSource
+
+func (pss v4PermissionSources) IsViaOrg(combined Permission) bool {
```
I'm not a fan of this function; will probably revisit tomorrow and might switch it to make a few more API requests to avoid assuming quite so much about how the API will behave in the future.

At minimum, it needs a full set of tests for the described functionality.

bluekeyes replied:
Replaced in latest update. I think the extra API requests (which should only be 1 in practice) make it easier to understand and less fragile to future API changes.

bluekeyes reviewed (commented):
It compiles and passes the existing tests now, but it still needs tests for the new functionality. I'm also not happy about how I worked around current GitHub API limits and may revisit that part.

bluekeyes committed:
Merge branch 'develop' into bkeyes/maintain-triage

bluekeyes committed:
Move some types around in pull package

bluekeyes committed:
Improve method to determine permission sources

Use additional API requests (probably 1 in practice) to get this
information more directly, instead of using heuristics based on API
misbehavior.

bluekeyes committed:
Simplify reviewer assignment tests

Only include results relevant to situation being tested to minimize
breaks when unrelated behavior changes.

bluekeyes committed:
Add tests for permission-based reviewer assignment

bluekeyes committed:
Add test for RepositoryCollaborators method

bluekeyes commented:
Still need to do some more local testing, but this should be ready for review now. This ended up including a bit of refactoring for reviewer assignment that wasn't strictly necessary for the new permission types, but should make things more efficient when using `teams` as the request mode.

bluekeyes committed:
Actually set ReviewRequestRule on results

Started refactoring this in a previous commit and then never finished.

bluekeyes commented on pull/github.go:
```diff
@@ -309,44 +310,170 @@ func (ghc *GitHubContext) Reviews() ([]*Review, error) {
+					PageInfo v4PageInfo
+					Edges    []struct {
+						Permission        string
+						PermissionSources []struct {
+							Source struct {
+								Team struct {
+									Slug string
+								} `graphql:"... on Team"`
+							}
+						}
```
This _should_ work, and seems to behave correctly with a personal access token, but I can't get `permissionSources` to resolve to a non-empty list when using a GitHub App token. Filed a ticket with support, but will probably have to replace this with direct (and more expensive) team listing to check membership.

bluekeyes replied:
Replaced with team member listing in the latest commit. Fortunately, I think we'd perform most these listings anyway in any cases where `RepositoryCollaborators` is currently used, so net requests probably aren't increasing too badly.

asvoboda commented on policy/reviewer/reviewer.go:
```diff
@@ -150,163 +149,146 @@ func selectOrgMembers(prctx pull.Context, allOrgs []string) ([]string, error) {
+	return nil
+}
+
+func selectUserReviewers(ctx context.Context, prctx pull.Context, selection *Selection, result *common.Result, r *rand.Rand) error {
```
this function feels a lot more readable now, thanks for fixing it up.

asvoboda commented on pull/permission.go:
```diff
@@ -0,0 +1,95 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pull
```
I wonder if it would be nice to put this in its own package (instead of `pull`), maybe it would be useful in octo-correct?

bluekeyes replied:
Given this is a pretty simple type and these two systems don't directly interact, I prefer copying this (or something similar) into other apps that need it rather than introducing a new dependency package or moving it into a common dependency like `go-githubapp` (where I feel it doesn't really fit with any existing functionality.)

asvoboda replied:
I was thinking maybe a package in policy-bot, but I think you're right.

asvoboda commented on pull/github.go:
```diff
@@ -309,44 +310,170 @@ func (ghc *GitHubContext) Reviews() ([]*Review, error) {
+	var p Permission
+	if len(q.Repository.Collaborators.Edges) > 0 {
+		var err error
+		p, err = ParsePermission(q.Repository.Collaborators.Edges[0].Permission)
```
What happens if we have more than one `edge`?

bluekeyes replied:
I was assuming this wouldn't happen because there's a query parameter for the user name and users must be unique. I guess I should check to make sure it doesn't do substring matches or something like that.

bluekeyes replied:
Good catch, the `query` parameter is a substring filter. Fixed in the latest commit to iterate through the results to make sure we find an exact match.

bluekeyes committed:
Remove excessive debug logging

bluekeyes committed:
Fix collaborator lookup for substring matches

I mistakenly assumed the "query" parameter was an exact match on user
name, but it's actually a substring filter. This means we need to
iterate over returned collaborators to find the one we want.

bluekeyes committed:
Fix collaborator listing for missing GraphQL field

permissionsSources is apparently not available to GitHub Apps,
regardless of the app's permissions.

bluekeyes committed:
Merge branch 'develop' into bkeyes/maintain-triage

bluekeyes commented:
Ok, this should be set for final review now. Local testing looks good and I think I found all the API scorpions 🦂.

asvoboda commented:
👍  Lets try it out!
